### ISSUE TITLE
Standard deviation parameters are incorrectly treated as variances in statespace covariance matrices

### ISSUE THREAD

**Issue by rklees** [pymc-devs/pymc-experimental#295]
I simulate data from a simple integrated random walk model defining the initial state, the measurement disturbance standard deviation (here, I set it equal to 0), and the trend disturbance standard deviation. Given these parameters, the simulated data should be close to a straight line. However, is looks completely different. Therefore, I used statsmodel to simulate the data using the same initial state and disturbance standard deviations. Also the random generator is initialized the same in both simulations. The simulate dataset from statsmodel is close to a straight line as expected. Therefore, I wonder whether there is something wrong in simulate_from_numpy_model. Here is the code:

measurement_error = st.MeasurementError(name="obs")
IRW = st.LevelTrendComponent(order=2, innovations_order=[0, 1])

param_dict = {
    "initial_trend": [0., 0.01], "sigma_trend": np.array([np.sqrt(5e-7)]),
    "sigma_obs": np.array([0.]),
}  

mod = IRW + measurement_error
nobs = 1024 # we simulate a time series of length Nobs

x, y = simulate_from_numpy_model(mod, np.random.default_rng(2163), param_dict, steps=nobs)

And here is the code from statsmodel:
model = sm.tsa.UnobservedComponents(y, level=True, stochastic_level=False, trend=True, stochastic_trend=True, irregular=True)
res = model.fit()
y_sim = model.simulate(initial_state = [0., 0.01], params=[0., 5e-7], nsimulations=nobs, random_state=np.random.default_rng(2163))

y and y_sim have nothing common. Whereas y_sim represents a slight deviation from the straight line, y is completely different, with a strong negative slope, though the undisturbed slope is positive. Attached the simulated data in both cases. 
![simulate_from_numpy_model](https://github.com/pymc-devs/pymc-experimental/assets/93590629/29aef3e3-0831-47f3-89ca-b2cb73c596da)
![statsmodel](https://github.com/pymc-devs/pymc-experimental/assets/93590629/93d29622-424c-4fec-be86-86f4a37cbac0)

**Comment by ricardoV94** [comment]
CC @jessegrabowski

**Comment by jessegrabowski** [comment]
I can take a look, but this is just a debug function. Why are you using it?

**Comment by rklees** [comment]
I use it because it makes data generation for a given state-space model easy, much better than statsmodel (as far as I know).

**Comment by jessegrabowski** [comment]
In this specific case, statsmodels expects sigma squared, whereas the structural model expects sigma. If you remove the square root from the structural model (but not the statsmodel argument), you will get the same output, modulo random draws. I don't think setting the seed is enough to guarantee the same outputs between the two functions, because there might be differences in how/when each random number is generated (all at once vs one-by-one in loop, for example).

I'm not sure what your exact use-case in, but in general I suggest you use the usual PyMC API to generate draws from the statespace models, rather than the testing function.

**Comment by rklees** [comment]
I know that statsmodel expects variances and pymc-experimental standard deviations. That is not the problem. The dataset pymc- experimental generates given the midel parameters and initial state is completely wrong. I can make a gist so that you can run my code yourself.
R Klees

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by jessegrabowski** [comment]
```
measurement_error = st.MeasurementError(name="obs")
IRW = st.LevelTrendComponent(order=2, innovations_order=[0, 1])

# Change sigma_trend to remove the square root
param_dict = {
"initial_trend": [0., 0.01], "sigma_trend": np.array([5e-7]),
"sigma_obs": np.array([0.]),
}

mod = IRW + measurement_error
nobs = 1024 # we simulate a time series of length Nobs

x, y = simulate_from_numpy_model(mod, np.random.default_rng(2163), param_dict, steps=nobs)
plt.plot(y)
# sm part unchanged 
```
![Untitled](https://github.com/pymc-devs/pymc-experimental/assets/48652735/b5a99ff1-e888-4b2b-92b9-5af9f66e9571)

This now looks correct to me?

**Comment by rklees** [comment]
The plot I posted was generated with:
measurement_error = st.MeasurementError(name="obs")
IRW = st.LevelTrendComponent(order=2, innovations_order=[0, 1])
mod = IRW + measurement_error
param_dict = {
    "initial_trend": np.array([0., 0.01]), "sigma_trend": np.array([np.sqrt(5e-7)]),
    "sigma_obs": np.array([0.]),
}
nobs = 1024
x, y = simulate_from_numpy_model(mod, np.random.default_rng(2163), param_dict, steps=nobs)

Because sigma_trend is the slope disturbance standard deviation as the name says. Hence, I assumed that in param_dict I have to input the standard deviations of the state disturbance terms. But obviously not? I noticed that for the IRW model, the matrix Q = [sigma_trend], but should be [sigma_trend*signa_trend], as it is the state covariance matrix.

So, please tell me whether in param_dict I have to input the state disturbance variances (then, the name should be sigma2_trend etc.) and not the state disturbance standard deviations (then, the name should be sigma_trend etc).

R Klees

From: Jesse Grabowski ***@***.***>
Reply to: pymc-devs/pymc-experimental ***@***.***>
Date: Thursday, 4 January 2024 at 18:29
To: pymc-devs/pymc-experimental ***@***.***>
Cc: Roland Klees ***@***.***>, Author ***@***.***>
Subject: Re: [pymc-devs/pymc-experimental] Something wrong with simulate_from_numpy_model ?? (Issue #295)


measurement_error = st.MeasurementError(name="obs")

IRW = st.LevelTrendComponent(order=2, innovations_order=[0, 1])



# Change sigma_trend to remove the square root

param_dict = {

"initial_trend": [0., 0.01], "sigma_trend": np.array([5e-7]),

"sigma_obs": np.array([0.]),

}



mod = IRW + measurement_error

nobs = 1024 # we simulate a time series of length Nobs



x, y = simulate_from_numpy_model(mod, np.random.default_rng(2163), param_dict, steps=nobs)

plt.plot(y)

# sm part unchanged

Untitled.png (view on web)<https://urldefense.com/v3/__https:/github.com/pymc-devs/pymc-experimental/assets/48652735/b5a99ff1-e888-4b2b-92b9-5af9f66e9571__;!!PAKc-5URQlI!97wysGBjWmVV157W3XU9OlmsBrPL8m2yJdsPpvwNGI7aFHTOJhKLqk_w2kvqOQTcpfkyAnBkvECDcNE_X-bWvbKyIQ$>

This now looks correct to me?

**Comment by jessegrabowski** [comment]
Yeah that's a fair point, the same is not consistent with what you get. Currently, whatever number you pass in is directly plugged into the Q matrix at whatever position on the main diagonal, so either it should be squared internally (this is what statsmodels does) or the name should be changed.

My preference would be to square things internally and keep the name the same, because no other PyMC distributions are parameterized by the variance directly. Does that seem reasonable?

**Comment by rklees** [comment]
I agree. In fact, it does not matter as long as the names are consistent. As is, they are not.
Roland

From: Jesse Grabowski ***@***.***>
Reply to: pymc-devs/pymc-experimental ***@***.***>
Date: Thursday, 4 January 2024 at 18:44
To: pymc-devs/pymc-experimental ***@***.***>
Cc: Roland Klees ***@***.***>, Author ***@***.***>
Subject: Re: [pymc-devs/pymc-experimental] Something wrong with simulate_from_numpy_model ?? (Issue #295)


Yeah that's a fair point, the name is not consistent with what you get. Currently, whatever number you pass in is directly plugged into the Q matrix at whatever position on the main diagonal, so either it should be squared internally (this is what statsmodels does) or the name should be changed.

My preference would be to square things internally and keep the name the same, because no other PyMC distributions are parameterized by the variance directly. Does that seem reasonable?

**Comment by rklees** [comment]
By the way, same applies to “sigma_obs”
Roland

From: Jesse Grabowski ***@***.***>
Reply to: pymc-devs/pymc-experimental ***@***.***>
Date: Thursday, 4 January 2024 at 18:44
To: pymc-devs/pymc-experimental ***@***.***>
Cc: Roland Klees ***@***.***>, Author ***@***.***>
Subject: Re: [pymc-devs/pymc-experimental] Something wrong with simulate_from_numpy_model ?? (Issue #295)


Yeah that's a fair point, the name is not consistent with what you get. Currently, whatever number you pass in is directly plugged into the Q matrix at whatever position on the main diagonal, so either it should be squared internally (this is what statsmodels does) or the name should be changed.

My preference would be to square things internally and keep the name the same, because no other PyMC distributions are parameterized by the variance directly. Does that seem reasonable?

**Comment by jessegrabowski** [comment]
It applies to everything in every structural model

**Comment by rklees** [comment]
Dear Jesse,

Do I only need to download the latest version of pymc_experimental to fix this bug?

Roland Klees

From: Jesse Grabowski ***@***.***>
Reply to: pymc-devs/pymc-experimental ***@***.***>
Date: Sunday, 7 January 2024 at 01:49
To: pymc-devs/pymc-experimental ***@***.***>
Cc: Roland Klees ***@***.***>, Author ***@***.***>
Subject: Re: [pymc-devs/pymc-experimental] Standard deviation parameters are incorrectly treated as variances in statespace covariance matrices (Issue #295)


Closed #295<https://urldefense.com/v3/__https:/github.com/pymc-devs/pymc-experimental/issues/295__;!!PAKc-5URQlI!-G6k2-DygvWZRuIXSL4rymvpTCII2Evf2BPl8bqNDE77VnoU_Tn-aSMVkUsZbKyxnoDyfGKFcPk0rxR3B5qs4bvVvQ$> as completed via #296<https://urldefense.com/v3/__https:/github.com/pymc-devs/pymc-experimental/pull/296__;!!PAKc-5URQlI!-G6k2-DygvWZRuIXSL4rymvpTCII2Evf2BPl8bqNDE77VnoU_Tn-aSMVkUsZbKyxnoDyfGKFcPk0rxR3B5odPse4_A$>.

**Comment by jessegrabowski** [comment]
You'll have to pip install the main branch from github. We haven't done a new release yet, so just doing `pip install pymc-experimental` won't have the patch

**Comment by rklees** [comment]
I am not very familiar with github. What is the syntax which does this?
Roland

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by jessegrabowski** [comment]
It's similar to what you did before when you were installing from my branch, but now just targeting the main branch. Should be something like:

`pip install git+https://github.com/pymc-devs/pymc-experimental.git`

**Comment by rklees** [comment]
Sorry, but this command does not implement the changes. Everything stays the same.
Roland

From: Jesse Grabowski ***@***.***>
Reply to: pymc-devs/pymc-experimental ***@***.***>
Date: Sunday, 7 January 2024 at 19:41
To: pymc-devs/pymc-experimental ***@***.***>
Cc: Roland Klees ***@***.***>, Author ***@***.***>
Subject: Re: [pymc-devs/pymc-experimental] Standard deviation parameters are incorrectly treated as variances in statespace covariance matrices (Issue #295)


It's similar to what you did before when you were installing from my branch, but now just targeting the main branch. Should be something like:

pip install git+https://github.com/pymc-devs/pymc-experimental.git

**Comment by jessegrabowski** [comment]
Make sure you pip uninstall pymc-experimental first

**Comment by rklees** [comment]
Did it, and then used your syntax to install pymc_experimental. However, sigma_xxx is still the entry in Q and H.
Roland

From: Jesse Grabowski ***@***.***>
Reply to: pymc-devs/pymc-experimental ***@***.***>
Date: Sunday, 7 January 2024 at 20:04
To: pymc-devs/pymc-experimental ***@***.***>
Cc: Roland Klees ***@***.***>, Author ***@***.***>
Subject: Re: [pymc-devs/pymc-experimental] Standard deviation parameters are incorrectly treated as variances in statespace covariance matrices (Issue #295)


Make sure you pip uninstall pymc-experimental first

**Comment by rklees** [comment]
Your command just reinstalls pymc-experimental-0.0.16, which according to you does not contain the bug fix.
Roland

From: Jesse Grabowski ***@***.***>
Reply to: pymc-devs/pymc-experimental ***@***.***>
Date: Sunday, 7 January 2024 at 20:04
To: pymc-devs/pymc-experimental ***@***.***>
Cc: Roland Klees ***@***.***>, Author ***@***.***>
Subject: Re: [pymc-devs/pymc-experimental] Standard deviation parameters are incorrectly treated as variances in statespace covariance matrices (Issue #295)


Make sure you pip uninstall pymc-experimental first

**Comment by jessegrabowski** [comment]
I just tested the command on my computer and it seems like it works. There hasn't been a release, so the version won't be bumped by this command.

You can check the source code by e.g. `st.MeasurementError??` (in a Jupyter notebook) and check that you see the sigma being squared in the `make_symbolic_graph` method, like this:

```
    def make_symbolic_graph(self) -> None:
        error_sigma = self.make_and_register_variable(f"sigma_{self.name}", shape=(self.k_endog,))
        diag_idx = np.diag_indices(self.k_endog)
        idx = np.s_["obs_cov", diag_idx[0], diag_idx[1]]
        self.ssm[idx] = error_sigma**2
```

**Comment by rklees** [comment]
You are right. I had to restart the kernel to activate the changes.
Thanks a lot.

Roland

From: Jesse Grabowski ***@***.***>
Reply to: pymc-devs/pymc-experimental ***@***.***>
Date: Sunday, 7 January 2024 at 20:29
To: pymc-devs/pymc-experimental ***@***.***>
Cc: Roland Klees ***@***.***>, Author ***@***.***>
Subject: Re: [pymc-devs/pymc-experimental] Standard deviation parameters are incorrectly treated as variances in statespace covariance matrices (Issue #295)


I just tested the command on my computer and it seems like it works. There hasn't been a release, so the version won't be bumped by this command.

You can check the source code by e.g. st.MeasurementError?? (in a Jupyter notebook) and check that you see the sigma being squared in the make_symbolic_graph method, like this:

    def make_symbolic_graph(self) -> None:

        error_sigma = self.make_and_register_variable(f"sigma_{self.name}", shape=(self.k_endog,))

        diag_idx = np.diag_indices(self.k_endog)

        idx = np.s_["obs_cov", diag_idx[0], diag_idx[1]]

        self.ssm[idx] = error_sigma**2

**Comment by rklees** [comment]
Is it correct that in pymc.model I still define priors for the disturbance (shock) standard deviations (and not for the disturbance (shock) variances)?
Roland

From: Jesse Grabowski ***@***.***>
Reply to: pymc-devs/pymc-experimental ***@***.***>
Date: Sunday, 7 January 2024 at 20:29
To: pymc-devs/pymc-experimental ***@***.***>
Cc: Roland Klees ***@***.***>, Author ***@***.***>
Subject: Re: [pymc-devs/pymc-experimental] Standard deviation parameters are incorrectly treated as variances in statespace covariance matrices (Issue #295)


I just tested the command on my computer and it seems like it works. There hasn't been a release, so the version won't be bumped by this command.

You can check the source code by e.g. st.MeasurementError?? (in a Jupyter notebook) and check that you see the sigma being squared in the make_symbolic_graph method, like this:

    def make_symbolic_graph(self) -> None:

        error_sigma = self.make_and_register_variable(f"sigma_{self.name}", shape=(self.k_endog,))

        diag_idx = np.diag_indices(self.k_endog)

        idx = np.s_["obs_cov", diag_idx[0], diag_idx[1]]

        self.ssm[idx] = error_sigma**2

**Comment by jessegrabowski** [comment]
Everything works as it appears now. You put a prior on sigma (the standard deviation), which is then squared and placed on the diagonal of the covariance matrix.

**Comment by rklees** [comment]
Great. Thanks.
Roland

From: Jesse Grabowski ***@***.***>
Reply to: pymc-devs/pymc-experimental ***@***.***>
Date: Sunday, 7 January 2024 at 20:56
To: pymc-devs/pymc-experimental ***@***.***>
Cc: Roland Klees ***@***.***>, Author ***@***.***>
Subject: Re: [pymc-devs/pymc-experimental] Standard deviation parameters are incorrectly treated as variances in statespace covariance matrices (Issue #295)


Everything works as it appears now. You put a prior on sigma (the standard deviation), which is then squared and placed on the diagonal of the covariance matrix.
